### PR TITLE
[amd64] Make code patching smp safe for the case where we emit a mov …

### PR DESCRIPTION
**Pull request by vargaz**
…+ call_reg. Hopefully fixes #57959.

### CODE REVIEW

**Comment by terrylewis60** [comment]
Please stop fucking sending shit you can see that I'm deleted fucking shit
are you dumb

On Aug 25, 2017 6:46 PM, "Zoltan Varga" <notifications@github.com> wrote:

…+ call_reg. Hopefully fixes #57959.
------------------------------
You can view, comment on, or merge this pull request online at:

  https://github.com/mono/mono/pull/5452
Commit Summary

   - [amd64] Make code patching smp safe for the case where we emit a mov +
   call_reg. Hopefully fixes #57959.

File Changes

   - *M* mono/mini/cpu-amd64.md
   <https://github.com/mono/mono/pull/5452/files#diff-0> (4)
   - *M* mono/mini/mini-amd64.c
   <https://github.com/mono/mono/pull/5452/files#diff-1> (4)

Patch Links:

   - https://github.com/mono/mono/pull/5452.patch
   - https://github.com/mono/mono/pull/5452.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/mono/mono/pull/5452>, or mute the thread

**Comment by terrylewis60** [comment]
Now stop before I report you

File: mono/mini/mini-amd64.c [hunk 1]
```diff
@@ -2870,6 +2870,10 @@ emit_call_body (MonoCompile *cfg, guint8 *code, MonoJumpInfoType patch_type, gco
 			amd64_call_code (code, 0);
 		}
 		else {
+			if (!no_patch && ((guint32)(code + 2 - cfg->native_code) % 4) != 0) {
+				guint32 pad_size = 4 - ((guint32)(code + 2 - cfg->native_code) % 4);
```
Comment (kumpera) [inline mono/mini/mini-amd64.c]: Shouldn't we align to 8 bytes here? Aligning at 4 we still risk cross a cache line

**Comment by kumpera** [comment]
LGTM, could explain how you found this issue? Everyone is curious about it!!

**Comment by lewurm** [comment]
that should probably be backported to at least `2017-08`

**Comment by marek-safar** [comment]
Yes, this needs to go to 2017-08, 2017-06 and d15-4-2017-06 branches

**Pull request merged**